src/Primitives/Search.purs

```diff
@@ -17,67 +22,115 @@ import Halogen.HTML.Properties as HP
 import Halogen.Query.HalogenM (fork, raise) as H
 import Select.Primitives.State (State, updateStore, getState)
 import Select.Primitives.Container as C
-import Select.Effects (Effects, FX)
-
-{-
-
-The Search primitive captures user input and returns it to the parent.
-
--}
+import Select.Effects (Effects)
 
 -- | The query type for the `Search` primitive. This primitive handles text input
--- | and debouncing.
+-- | and debouncing. It has a special query for the purpose of embedding Container
+-- | queries, used to route key events to the container from an input field.
+-- |
+-- | Arguments:
+-- |
+-- | - `o`: The type of the parent query to embed. This is usually the component that
+-- | -      mounts the search primitive, but it could also be the query type of some
+-- |        higher component.
+-- | - `item`: Your custom item type, used by your renderers.
+-- | - `eff`: The row of effects to use in the primitive. You are expected to provide your
+-- |          rows wrapped in the select Effects type:
+-- |
+-- | ```purescript
+-- | type ChildQuery eff = SearchQuery MyQuery MyItem (Effects eff)
+-- | ```
+-- |
+-- | Constructors:
 -- |
 -- | - `TextInput`: Handle new text input as a string
+-- | - `Raise`: Embed a parent query that can be returned to the parent for evaluation
+-- | - `FromContainer`: Embed a container query that can be routed to a container slot
 -- | - `SearchReceiver`: Update the component with new `Input` when the parent re-renders
-data SearchQuery o item e a
+data SearchQuery o item eff a
   = TextInput String a
   | Raise (o Unit) a
   | FromContainer (C.ContainerQuery o item Unit) a
-  | SearchReceiver (SearchInput o item e) a
+  | SearchReceiver (SearchInput o item eff) a
 
--- | The `Search` primitive internal state
+-- | The `Search` primitive internal state.
+-- |
+-- | Arguments:
+-- |
+-- | - `eff`: The row of effects used in the primivite. This needs to be wrapped in the
+-- |          effects type in practice:
+-- |
+-- | ```purescript
+-- | forall e. SearchState (Effects e)
+-- | ```
 -- |
--- | - `search`: The `String` contained within the primitive
+-- | Fields:
+-- |
+-- | - `search`: The `String` contained within the primitive. This is captured from user
+-- |             input, or can be set by the parent.
 -- | - `ms`: Number of milliseconds for the input to be debounced before passing
 -- |         a message to the parent. Set to 0.0 if you don't want debouncing.
--- | - `debouncer`: Used to facilitate debouncing of the input
-type SearchState e =
+-- | - `debouncer`: Facilitates debouncing for the input field.
+type SearchState eff =
   { search    :: String
   , ms        :: Milliseconds
-  , debouncer :: Maybe (Debouncer e)
+  , debouncer :: Maybe (Debouncer eff)
   }
 
 -- | The `Debouncer` type alias, used to debounce user input in the `Search` primitive.
-type Debouncer e =
+type Debouncer eff =
   { var   :: AVar Unit
-  , fiber :: Fiber (Effects e) Unit }
+  , fiber :: Fiber eff Unit }
```

Comment:
![](https://media.giphy.com/media/XsUtdIeJ0MWMo/giphy.gif)

Reply:
You would not believe...

Reply:
👍